File: week6/p2.py
class Node:
    def __init__(self, new_data):
        self.data = new_data
        self.next = None

class myQueue:
    def __init__(self):
        self.front = None
        self.rear = None
        self.count = 0

    def isEmpty(self):
        return self.count == 0

    def enqueue(self, x):
        newNode = Node(x)
        if self.rear is None:
            self.front = self.rear = newNode
        else:
            self.rear.next = newNode
            self.rear = newNode
        self.count += 1

    def dequeue(self):
        if self.isEmpty():
            return
        self.front = self.front.next
        if self.front is None:
            self.rear = None
        self.count -= 1

    def getFront(self):
        if self.isEmpty():
            return -1
        return self.front.data

    def size(self):
        return self.count

File: week6/p2.c
// Node structure
typedef struct Node {
    int data;
    struct Node* next;
} Node;

// Queue structure
typedef struct {
    Node* front;
    Node* rear;
    int currSize;
} myQueue;

// Function to create a new node
Node* createNode(int new_data) {
    Node* node = (Node*)malloc(sizeof(Node));
    node->data = new_data;
    node->next = NULL;
    return node;
}

// Initialize queue
void initQueue(myQueue* q) {
    q->front = q->rear = NULL;
    q->currSize = 0;
}

bool isEmpty(myQueue* q) {
    return q->currSize == 0;
}

void enqueue(myQueue* q, int x) {
    Node* newNode = createNode(x);
    if (q->rear == NULL) {
        q->front = q->rear = newNode;
    } else {
        q->rear->next = newNode;
        q->rear = newNode;
    }
    q->currSize++;
}

void dequeue(myQueue* q) {
    if (isEmpty(q)) return;
    Node* temp = q->front;
    q->front = q->front->next;
    if (q->front == NULL) q->rear = NULL;
    free(temp);
    q->currSize--;
}

int getFront(myQueue* q) {
    if (isEmpty(q)) return -1;
    return q->front->data;
}

int size(myQueue* q) {
    return q->currSize;
}